pkg/yurthub/poolcoordinator/resources/resources.go

```diff
@@ -0,0 +1,167 @@
+	poolScopeResources := make([]schema.GroupVersionResource, 0)
+	m.validPoolScopedResourcesLock.RLock()
+	defer m.validPoolScopedResourcesLock.RUnlock()
+	for _, v := range m.validPoolScopedResources {
+		poolScopeResources = append(poolScopeResources, v.GroupVersionResource)
+	}
+	return poolScopeResources
+}
+
+// addVerifiableGVRs add given gvrs to validPoolScopedResources map
+func (m *PoolScopeResourcesManger) addVerifiableGVRs(gvrs ...*verifiablePoolScopeResource) {
+	m.validPoolScopedResourcesLock.Lock()
+	defer m.validPoolScopedResourcesLock.Unlock()
+	for _, gvr := range gvrs {
+		if ok, errMsg := gvr.Verify(); ok {
+			m.validPoolScopedResources[gvr.String()] = gvr
+			klog.Infof("PoolScopeResourcesManger add gvr %s success", gvr.String())
+		} else {
+			klog.Warningf("PoolScopeResourcesManger add gvr %s failed, because %s", gvr.String(), errMsg)
+		}
+	}
+}
+
+// addVerifiableGVRs clear validPoolScopedResources and set given gvrs to validPoolScopedResources map
+func (m *PoolScopeResourcesManger) setVerifiableGVRs(gvrs ...*verifiablePoolScopeResource) {
+	m.validPoolScopedResourcesLock.Lock()
+	defer m.validPoolScopedResourcesLock.Unlock()
+	m.validPoolScopedResources = make(map[string]*verifiablePoolScopeResource)
+	for _, gvr := range gvrs {
+		if ok, errMsg := gvr.Verify(); ok {
+			m.validPoolScopedResources[gvr.String()] = gvr
+			klog.Infof("PoolScopeResourcesManger update gvr %s success", gvr.String())
+		} else {
+			klog.Warningf("PoolScopeResourcesManger update gvr %s failed, because %s", gvr.String(), errMsg)
+		}
+	}
+}
+
+func (m *PoolScopeResourcesManger) addConfigmap(obj interface{}) {
+	cfg, ok := obj.(*corev1.ConfigMap)
+	if !ok {
+		return
+	}
+
+	poolScopeResources := cfg.Data[util.PoolScopeResourcesKey]
+	poolScopeResourcesGVRs := make([]schema.GroupVersionResource, 0)
+	verifiablePoolScopeResourcesGVRs := make([]*verifiablePoolScopeResource, 0)
+	if err := json.Unmarshal([]byte(poolScopeResources), &poolScopeResourcesGVRs); err != nil {
+		klog.Errorf("PoolScopeResourcesManger unmarshal poolScopeResources %s failed with error = %s",
+			poolScopeResourcesGVRs, err.Error())
+		return
+	}
+	klog.Infof("PoolScopeResourcesManger add configured pool scope resources %v", poolScopeResourcesGVRs)
+	for _, v := range poolScopeResourcesGVRs {
+		verifiablePoolScopeResourcesGVRs = append(verifiablePoolScopeResourcesGVRs,
+			newVerifiablePoolScopeResource(v, m.getGroupVersionVerifyFunction(m.k8sClient)))
+	}
+	m.addVerifiableGVRs(verifiablePoolScopeResourcesGVRs...)
+}
+
+func (m *PoolScopeResourcesManger) getGroupVersionVerifyFunction(client kubernetes.Interface) func(gvr schema.GroupVersionResource) (bool, string) {
+	return func(gvr schema.GroupVersionResource) (bool, string) {
+		maxRetry := 3
+		duration := time.Second * 5
+		counter := 0
+		var err error
+		for counter <= maxRetry {
+			if _, err = client.Discovery().ServerResourcesForGroupVersion(gvr.GroupVersion().String()); err == nil {
```

Comment:
if cloud-edge network is off, client can not verify the gvr exist or not. maybe we need to store the gvr info into local disk for yurthub restart use.

Reply:
Yes, Shall we do that in the next release? At current release,  if cloud-edge network is off, the pool-coordinator would not work as there would be no leader. Let's do fix these issues together in the next release!
@rambohe-ch

Reply:
ok